src/main/java/io/jenkins/plugins/sshbuildagents/ssh/mina/SSHApacheMinaLauncher.java

```diff
@@ -0,0 +1,1125 @@
+        connection.connect();
+    }
+
+    /**
+     * Validates the Agent configuration.
+     *
+     * @throws InterruptedException
+     */
+    private void checkConfig() throws InterruptedException {
+        // JENKINS-58340 some plugins does not implement Descriptor
+        Descriptor descriptorOrg = Jenkins.get().getDescriptor(this.getClass());
+        if (!(descriptorOrg instanceof DescriptorImpl)) {
+            return;
+        }
+
+        DescriptorImpl descriptor = (DescriptorImpl) descriptorOrg;
+        String message = "Validate configuration:\n";
+        boolean isValid = true;
+
+        String port = String.valueOf(this.port);
+        FormValidation validatePort = descriptor.doCheckPort(port);
+        FormValidation validateHost = descriptor.doCheckHost(this.host);
+        FormValidation validateCredentials =
+                descriptor.doCheckCredentialsId(Jenkins.get(), Jenkins.get(), this.host, port, this.credentialsId);
+
+        if (validatePort.kind == FormValidation.Kind.ERROR) {
+            isValid = false;
+            message += validatePort.getMessage() + "\n";
+        }
+        if (validateHost.kind == FormValidation.Kind.ERROR) {
+            isValid = false;
+            message += validateHost.getMessage() + "\n";
+        }
+        if (validateCredentials.kind == FormValidation.Kind.ERROR) {
+            isValid = false;
+            message += validateCredentials.getMessage() + "\n";
+        }
+
+        if (!isValid) {
+            throw new InterruptedException(message);
+        }
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public void afterDisconnect(SlaveComputer slaveComputer, final TaskListener listener) {
+        if (connection == null) {
+            // Nothing to do here, the connection is not established
+            return;
+        } else {
+            try {
+                if (shellChannel != null) {
+                    if (shellChannel.getLastError() != null) {
+                        listener.getLogger()
+                                .println("\tException: "
+                                        + shellChannel.getLastError().getMessage());
+                    }
+                    if (StringUtils.isNotBlank(shellChannel.getLastHint())) {
+                        listener.getLogger().println("\tHint: " + shellChannel.getLastHint());
+                    }
+                }
+                close();
+            } catch (Exception e) {
+                listener.getLogger().println("Error after disconnect agent: " + e.getMessage());
+            }
+        }
+    }
+
+    /** Closes the SSH connection and any associated channels. */
+    private void close() {
+        try {
+            if (shellChannel != null) {
+                shellChannel.close();
+            }
+            if (connection != null) {
+                connection.close();
+            }
+        } catch (Exception e) {
```

Comment:
If `shellChannel.close()` throws an exception, should we still be trying to close `connection`?